spec/notification_center_spec.rb

```diff
@@ -0,0 +1,484 @@
+  end
+  
+  describe '#Notification center' do
+    
+    describe 'test add notification with invalid params' do
+      it 'should log and return nil if notification type is empty' do
+        expect(notification_center.add_notification_listener(
+         nil,
+         @callback_reference
+        )).to eq(nil)
+        expect(spy_logger).to have_received(:log).once
+                               .with(Logger::ERROR, 'Notification type can not be empty.')
+      end
+  
+      it 'should log and return nil if notification callback is empty' do
+        expect(notification_center.add_notification_listener(
+         Optimizely::NotificationCenter::NOTIFICATION_TYPES[:ACTIVATE],
+         nil
+        )).to eq(nil)
+        expect(spy_logger).to have_received(:log).once
+                               .with(Logger::ERROR, 'Callback can not be empty.')
+      end
+      
+      it 'should log and return nil if invalid notification type given' do
+        expect(notification_center.add_notification_listener(
+         'Test notification type',
+         @callback_reference
+        )).to eq(nil)
+        expect(spy_logger).to have_received(:log).once
+                               .with(Logger::ERROR, 'Invalid notification type.')
+      end
+      
+      it 'should log and return nil if invalid callable given' do
+        expect(notification_center.add_notification_listener(
+         Optimizely::NotificationCenter::NOTIFICATION_TYPES[:ACTIVATE],
+         'Invalid callback!'
+        )).to eq(nil)
+        expect(spy_logger).to have_received(:log).once
+           .with(Logger::ERROR, 'Invalid notification callback given.')
+      end
+    end
+    
+    describe 'test add notification with valid type and callback' do
+      it 'should add, and return notification ID when a plain function is passed as an argument ' do
+        expect(notification_center.add_notification_listener(
+         Optimizely::NotificationCenter::NOTIFICATION_TYPES[:ACTIVATE],
+         @callback_reference
+        )).to eq(1)
+        # verifies that one notification is added
+        expect(notification_center.notifications[Optimizely::NotificationCenter::NOTIFICATION_TYPES[:ACTIVATE]].length)
+         .to eq(1)
+      end
+    end
+    
+    describe 'test add notification for multiple notification types' do
+      it 'should add, return notification ID and validate callbacks when a valid callback is added for each notification type ' do
+        Optimizely::NotificationCenter::NOTIFICATION_TYPES.values.each_with_index {|value, index|
+          notification_id = index+1
+          expect(notification_center.add_notification_listener(
+           value,
+           @callback_reference
+          )).to eq(notification_id)
+        }
+        notification_center.notifications.each do |key, _array|
+          notification_center.notifications[key].each do |notification|
+            expect(notification[:callback]).to eq @callback_reference
```

Comment:
Why are you doing this

>  notification_center.notifications.each do |key, _array|
>  notification_center.notifications[key].each do |notification|
>  expect(notification[:callback]).to eq @callback_reference

Instead, you should iterate over notifications length and verify that length of each notifications type is now 1.